File: keboola_eshop/kbc.py
from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Any

import requests


@dataclass(frozen=True)
class KbcConfig:
    url: str
    token: str


class KbcError(RuntimeError):
    pass


def load_kbc_config() -> KbcConfig:
    url = (os.environ.get("KBC_URL") or "").strip().rstrip("/")
    token = (os.environ.get("KBC_TOKEN") or "").strip()

    if not url:
        raise KbcError("Missing environment variable: KBC_URL")
    if not token:
        raise KbcError("Missing environment variable: KBC_TOKEN")

    return KbcConfig(url=url, token=token)


def _headers(cfg: KbcConfig) -> dict[str, str]:
    return {
        "X-StorageApi-Token": cfg.token,
        "Accept": "application/json",
        "User-Agent": "keboola-eshop-local/0.1.0",
    }


def kbc_get(cfg: KbcConfig, path: str, *, params: dict[str, Any] | None = None) -> Any:
    path = path if path.startswith("/") else f"/{path}"
    url = f"{cfg.url}{path}"

    try:
        resp = requests.get(url, headers=_headers(cfg), params=params, timeout=30)
    except requests.RequestException as e:
        raise KbcError(f"Keboola request failed: {e}") from e

    if resp.status_code >= 400:
        body_preview = resp.text[:500]
        raise KbcError(f"Keboola API error {resp.status_code} for {path}: {body_preview}")

    if not resp.content:
        return None

    try:
        return resp.json()
    except ValueError as e:
        raise KbcError(f"Non-JSON response from Keboola for {path}") from e


def list_buckets(cfg: KbcConfig) -> Any:
    return kbc_get(cfg, "/v2/storage/buckets")


def list_tables(cfg: KbcConfig) -> Any:
    return kbc_get(cfg, "/v2/storage/tables")

File: src/kbc.js
import { readFileSync } from "node:fs";

export class KbcError extends Error {
  constructor(message, { status, bodyPreview } = {}) {
    super(message);
    this.name = "KbcError";
    this.status = status;
    this.bodyPreview = bodyPreview;
  }
}

export function loadEnvExplicit() {
  // Avoid dotenv's stack-based discovery; be explicit and predictable.
  try {
    readFileSync(".env");
    return true;
  } catch {
    return false;
  }
}

export function getKbcConfig() {
  const url = (process.env.KBC_URL || "").trim().replace(/\/+$/, "");
  const token = (process.env.KBC_TOKEN || "").trim();

  if (!url) throw new KbcError("Missing environment variable: KBC_URL");
  if (!token) throw new KbcError("Missing environment variable: KBC_TOKEN");

  return { url, token };
}

export async function kbcGet(path, { params } = {}) {
  const { url, token } = getKbcConfig();
  const fullPath = path.startsWith("/") ? path : `/${path}`;
  const u = new URL(`${url}${fullPath}`);
  if (params) {
    for (const [k, v] of Object.entries(params)) {
      if (v === undefined || v === null) continue;
      u.searchParams.set(k, String(v));
    }
  }

  let res;
  try {
    res = await fetch(u, {
      method: "GET",
      headers: {
        "X-StorageApi-Token": token,
        Accept: "application/json",
        "User-Agent": "keboola-eshop-local/0.1.0"
      }
    });
  } catch (e) {
    throw new KbcError(`Keboola request failed: ${e?.message || String(e)}`);
  }

  const text = await res.text();
  if (!res.ok) {
    throw new KbcError(`Keboola API error ${res.status} for ${fullPath}`, {
      status: res.status,
      bodyPreview: text.slice(0, 500)
    });
  }

  if (!text) return null;
  try {
    return JSON.parse(text);
  } catch {
    throw new KbcError(`Non-JSON response from Keboola for ${fullPath}`, {
      status: res.status,
      bodyPreview: text.slice(0, 500)
    });
  }
}

export function listBuckets() {
  return kbcGet("/v2/storage/buckets");
}

export function listTables() {
  return kbcGet("/v2/storage/tables");
}

